app/models/manageiq/providers/lenovo/physical_infra_manager/operations/ansible_sender.rb

```diff
@@ -0,0 +1,98 @@
+#
+# Mixin to run ansible operations
+#
+module ManageIQ::Providers::Lenovo::PhysicalInfraManager::Operations::AnsibleSender
+  extend ActiveSupport::Concern
+
+  #
+  # Executes ansible resources inside +root_dir+
+  #
+  # @param args [Hash] - The args to run the ansible resource.
+  # @option args [String] 'playbook_name' - The name of the playbook to be executed.
+  #         If this param is filled a yaml file that matches with the specified name
+  #         will be searched inside +playbooks_dir+.
+  #         In cases where the playbook is not in +playbooks_dir+ root, please, specify
+  #         the relative path to it.
+  #         Example:
+  #           run_ansible({'playbook_name' => 'config/firmware_update_all_dev'})
+  #
+  # @option args [String] 'role_name' - The name of the role to be executed.
+  #        If this param is filled it will search the role inside +roles_dir+.
+  #        Example:
+  #           run_ansible({'role_name' => 'lenovo.lxca-inventory'})
+  #
+  # @option args [Hash] vars - The set of vars needed to execute the ansible resource
+  #
+  # @return Ansible::Runner::Response
+  #
+  # @see #root_dir
+  # @see #playbooks_dir
+  # @see #roles_dir
+  #
+  def run_ansible(args = {})
+    playbook_name, role_name, vars = load_params(args)
+    if playbook_name.present?
```

Comment:
I really don't like methods that allow callers to throw anything at them and they figure it out from there.
Much prefer different methods for `.run` vs `.run_role`, is an abstraction layer even needed in that case?  Seems like you could just call `Ansible::Runner` directly.

Reply:
I see. Well the abstraction that this method brings is add default variables and playbooks directory to `Ansible::Runner` calls. Nothing more fair that the provider abstract these information, once the playbook dir lives in `manageiq-providers` project and it owns the provider vars. 

The `run` and `run_role` methods already exists on `Ansible::Runner` so this method also adds an abstraction on it, as well. This way, this method will be useful when we implement more specific methods, like `ansible_update_firmware` or `ansible_provisioning_server` for example, where they will just call the same method, changing only the variables. Beside that it is being already useful in our [Api implementation](https://github.com/ManageIQ/manageiq-api/pull/474) where it allow clients to call any ansible resource (playbook or role) in a common way, making easier to implement ansible based buttons on UI.

Reply:
> I see. Well the abstraction that this method brings is add default variables and playbooks directory to Ansible::Runner calls. Nothing more fair that the provider abstract these information, once the playbook dir lives in manageiq-providers project and it owns the provider vars.

I don't have any issue with the provider having some default variables

> The run and run_role methods already exists on Ansible::Runner so this method also adds an abstraction on it, as well.

This is what I have an issue with, a method that can accept any combination of arguments that cause it to do different things is harder to debug later and harder to assert the parameters are valid.

If you want to have one method to do it all that's up to you

Reply:
> This is what I have an issue with, a method that can accept any combination of arguments that cause it to do different things is harder to debug later and harder to assert the parameters are valid.

I see your point, and I agree. But, since it is a decision that have to be made at some point (API call or `run_ansible` call) we decide, at least for now, leave this logic here. Thanks for your tips, I'll try to figure out a better implementation for it